tests/unit/s2n_early_data_test.c

```diff
@@ -312,5 +326,50 @@ int main(int argc, char **argv)
         EXPECT_EQUAL(psk->early_data_config.context.allocated, 0);
     }
 
+    /* Test s2n_early_data_config_clone */
+    {
+        const uint8_t test_bad_value[] = "wrong";
+        const uint8_t test_apln[] = "protocol";
+        const uint8_t test_context[] = "context";
+        const uint32_t test_max_early_data = 10;
+        const struct s2n_cipher_suite *test_cipher_suite = &s2n_tls13_chacha20_poly1305_sha256;
+
+        for (size_t called_directly = 0; called_directly < 2; called_directly++) {
+            struct s2n_psk *original = s2n_external_psk_new();
+            EXPECT_NOT_NULL(original);
+            EXPECT_SUCCESS(s2n_psk_configure_early_data(original, test_max_early_data,
+                    test_cipher_suite->iana_value[0], test_cipher_suite->iana_value[1]));
+            EXPECT_SUCCESS(s2n_psk_set_application_protocol(original, test_apln, sizeof(test_apln)));
+            EXPECT_SUCCESS(s2n_psk_set_context(original, test_context, sizeof(test_context)));
+
+            DEFER_CLEANUP(struct s2n_psk *clone = s2n_external_psk_new(), s2n_psk_free);
+            EXPECT_SUCCESS(s2n_psk_set_application_protocol(clone, test_bad_value, sizeof(test_bad_value)));
+            EXPECT_NOT_NULL(clone);
+
+            if (called_directly) {
+                EXPECT_OK(s2n_early_data_config_clone(clone, &original->early_data_config));
+            } else {
+                EXPECT_SUCCESS(s2n_psk_set_identity(original, test_bad_value, sizeof(test_bad_value)));
+                EXPECT_SUCCESS(s2n_psk_set_secret(original, test_bad_value, sizeof(test_bad_value)));
+                EXPECT_OK(s2n_psk_clone(clone, original));
+            }
+
+            /* Free the original to ensure they share no memory */
+            EXPECT_SUCCESS(s2n_psk_free(&original));
```

Comment:
I don't like this as a forcing function as it depends on three other pieces of the system working correctly:
1. Calling `s2n_psk_free` zeroizes the fields we're checking. (It appears to zeroize `early_data_config`, so we should be safe)
2. These values we are testing against aren't set to `zero`
3. `original` doesn't drag references and thus won't accidentally zeroize the data we're checking against.

All of these happen to be safe, but they are all implicit assumptions and some are easier to accidentally invalidate than others.

Prior to calling `s2n_psk_free(&original)` could we also check to ensure some of the internal pointers of `original` and `clone` are different? This would be sufficient to make me happy.